core/schema/git.go

```diff
@@ -120,62 +125,141 @@ func (s *gitSchema) git(ctx context.Context, parent *core.Query, args gitArgs) (
 					},
 				},
-			); err != nil {
-				return nil, fmt.Errorf("failed to select internal socket: %w", err)
-			}
+			},
+		); err != nil {
+			return nil, fmt.Errorf("failed to select internal socket: %w", err)
+		}
 
-			if err := socketStore.AddUnixSocket(sockInst.Self, clientMetadata.ClientID, clientMetadata.SSHAuthSocketPath); err != nil {
-				return nil, fmt.Errorf("failed to add unix socket to store: %w", err)
-			}
-			authSock = sockInst.Self
+		if err := socketStore.AddUnixSocket(sockInst.Self, clientMetadata.ClientID, clientMetadata.SSHAuthSocketPath); err != nil {
+			return nil, fmt.Errorf("failed to add unix socket to store: %w", err)
 		}
+		authSock = sockInst.Self
 	}
 
+	// 4. Handle git directory configuration
 	discardGitDir := false
 	if args.KeepGitDir != nil {
 		slog.Warn("The 'keepGitDir' argument is deprecated. Use `tree`'s `discardGitDir' instead.")
 		discardGitDir = !*args.KeepGitDir
 	}
 
+	// 5. Handle PAT auth
+	var authToken *core.Secret = nil
+	mainClientCallerID, err := parent.Server.MainClientCallerID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve mainClientCallerID: %w", err)
+	}
+
+	// Isolate auth token to main client
+	if clientMetadata != nil && clientMetadata.ClientID == mainClientCallerID {
+		// extract scheme from ref
+		remote, err := gitutil.ParseURL(args.URL)
+		if errors.Is(err, gitutil.ErrUnknownProtocol) {
+			remote, err = gitutil.ParseURL("https://" + args.URL)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse Git URL: %w", err)
+		}
+
+		if remote.Scheme == "https" || remote.Scheme == "http" {
+			// Check if repo is public
+			repo := git.NewRemote(memory.NewStorage(), &config.RemoteConfig{
+				Name: "origin",
+				URLs: []string{remote.Remote},
+			})
+
+			_, err := repo.List(&git.ListOptions{Auth: nil})
+			if err == nil {
+				// Repository is public, no need for auth token
+				return &core.GitRepository{
+					Query:         parent,
+					URL:           args.URL,
+					DiscardGitDir: discardGitDir,
+					SSHKnownHosts: args.SSHKnownHosts,
+					SSHAuthSocket: authSock,
+					Services:      svcs,
+					Platform:      parent.Platform(),
+					AuthToken:     nil,
+				}, nil
+			}
+
+			// Only proceed with auth if repo requires authentication
+			if errors.Is(err, transport.ErrAuthenticationRequired) {
+				bk, err := parent.Buildkit(ctx)
+				if err != nil {
+					return nil, fmt.Errorf("failed to get buildkit: %w", err)
+				}
+
+				// Retrieve credential from host
+				credentials, err := bk.GetCredential(ctx, remote.Scheme, remote.Host, remote.Path)
+				if err == nil {
+					// Credentials found, create and set auth token
+					var secretAuthToken dagql.Instance[*core.Secret]
+					if err := s.srv.Select(ctx, s.srv.Root(), &secretAuthToken,
+						dagql.Selector{
+							Field: "setSecret",
+							Args: []dagql.NamedInput{
+								{
+									Name:  "name",
+									Value: dagql.NewString("gitAuthtoken"),
```

Comment:
Not sure if this is still the case, but iirc these secret names are session-wide. So if you use two separate private repos in the same session it seems like these might clobber. Maybe just set a name that's a hash of the plaintext, like we do for the CLI?

Reply:
@sipsma, are these session-wide ?

Reply:
Secrets are stored per-client so that sort of clobbering shouldn't be an issue.

The only extremely obscure case I can think of that would be impacted is:
1. You're using an SDK directly on the host (i.e. not just `dagger call`)
2. You set your own secret that also happens to be named `gitAuthtoken`

Not worth agonizing over, but yeah we could just set name to a sha256 of the plaintext to squash any possibility of that happening.